### ISSUE TITLE
Create "price concession" measure for website

### ISSUE THREAD

**Issue by richiecroker** [ebmdatalab/openprescribing#1179]
During the price concession issues last year, we created a series of blogs showing an analysis of the impact on each CCG.

The analysis included an estimate based on the latest month's data for current concessions.  The SQL used was:

```sql
SELECT
  month,
  rx.pct,
  rx.bnf_name,
  rx.bnf_code,
  dt_price_pence,
  price_concession_pence,
  qtyval AS packsize,
  SUM(quantity/qtyval) AS no_of_packs,
  SUM(quantity) AS quantity,
  SUM(quantity*ncso.add_cost_per_qty) AS add_cost
FROM
  `hscic.normalised_prescribing_standard` AS rx
JOIN
  `dmd.ncso_calculator` AS ncso
ON
  rx.bnf_code = ncso.bnf_code
  AND rx.month=ncso.date
JOIN
  hscic.ccgs AS ccg
ON
  rx.pct=ccg.code
GROUP BY
  rx.pct,
  rx.bnf_code,
  rx.bnf_name,
  month,
  ncso.add_cost_per_qty,
  dt_price_pence,
  price_concession_pence,
  qtyval
UNION ALL
SELECT
  ncso.date,
  rx.pct,
  rx.bnf_name,
  rx.bnf_code,
  dt_price_pence,
  price_concession_pence,
  qtyval AS packsize,
  SUM(quantity/qtyval) AS no_of_packs,
  SUM(quantity) AS quantity,
  SUM(quantity*ncso.add_cost_per_qty) AS add_cost
FROM
  `hscic.normalised_prescribing_standard_latest_month` AS rx
JOIN
  `dmd.ncso_calculator` AS ncso
ON
  rx.bnf_code = ncso.bnf_code
JOIN
  hscic.ccgs AS ccg
ON
  rx.pct=ccg.code
WHERE
  ncso.date>(
  SELECT
    MAX(rx1.month)
  FROM
    hscic.normalised_prescribing_standard_latest_month AS rx1)
GROUP BY
  rx.pct,
  rx.bnf_code,
  rx.bnf_name,
  ncso.date,
  ncso.add_cost_per_qty,
  dt_price_pence,
  price_concession_pence,
  qtyval
ORDER BY
  month,
  bnf_code,
  pct
```

As there are more price concession issues appearing at the moment, it would be good to recreate this analysis on the website, using a page similar to this:

![image](https://user-images.githubusercontent.com/18683239/48348291-d93e4180-e678-11e8-9452-8a37870a78f7.png)

**Comment by richiecroker** [comment]
We will need to consider how we deal with where there have been price concessions on one pack size where multiple pack sizes exist, although this is very uncommon (once with trimethoprim, once with tramadol).

**Comment by sebbacon** [comment]
Regarding the multiple pack size problem, BSA informed @richiecroker by email:

> I can confirm that Trimethoprim 200mg tablets had only the 6 tablet pack size on the concession price list in June 2018.
> If an item is in the Drug Tariff and there are multiple packs listed, we will go with the contractor's endorsement of one of the pack sizes.
>
> If there was no endorsement of a pack size, the following would occur:
> •    If a pack size exists for the exact quantity prescribed, payment will be made for that pack.
> •    If the quantity prescribed is an exact multiple of a pack listed in the tariff then we will pay a price based on that.
> •    If there is no matching pack size to the quantity prescribed, payment will be made based on supply from the next largest pack size to the quantity prescribed.
> It is The Department of Health and Social Care that decides to grant a concessionary price in addition to the relevant pack sizes. The NHS Business Services Authority is not involved in the decision.

As we have no way of knowing if a pack is endorsed or not, there are two ways of approaching this problem:

1. Where there has previously been a split price-per-unit within NCSO for the same presentation for that CCG, work out the mean price-per-unit in a previous month, and then adjust the figure for projected months accordingly
2. Use the cheapest price-per-unit for calculations, and describe this as "best case"
3. Use the most expensive price-per-unit, and describe this as "worst case"

I'm tempted to go for (2).

**Comment by sebbacon** [comment]
I think we should go for the implementation @rich has drawn up in the first instance. HTML/JS-wise we can use the same datatable approach as for [measure breakdowns](https://github.com/ebmdatalab/openprescribing/blob/ddebc66117f8a1840e5769ae7cd6b218436acb00/openprescribing/templates/_measure_breakdown_table.html).  Generating the data from bigquery would use some of the logic we already have in the pipeline (see gcutils/bigquery.py et al); download to local model; serve data over API; profit.

However, a couple of other things that have occurred to me while thinking about this:

1. Could this actually be implemented as a measure? Then there's no new data models, and you get national deciles for context, and the contributing presentations table/download is already implemented. The problem is we're projecting 3 months into the future which would touch various bits of code. Another benefit is we could apply the same projection logic to all (cost-based) measures

2. It might be interesting to include all spending, not just on NCSO items, as we have all the DT prices up to the current month to play with. This would be a stacked bar chart showing standard spending + NCSO spending = total spending.  @richiecroker says he'd have to see it to be able to say if it would be adding a lot!

**Comment by evansd** [comment]
Closes in #1249
